src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -894,72 +893,59 @@ static BOOL AttributeHasChangesOmittedQuality(MTRAttributePath * attributePath)
 
         // Create work item, set ready handler to perform task, then enqueue the work
         MTRAsyncWorkItem * workItem = [[MTRAsyncWorkItem alloc] initWithQueue:self.queue];
-        [workItem setBatchingID:MTRDeviceWorkItemBatchingReadID data:readRequests handler:^(id opaqueDataCurrent, id opaqueDataNext, BOOL * fullyMerged) {
+        [workItem setBatchingID:MTRDeviceWorkItemBatchingReadID data:readRequests handler:^(id opaqueDataCurrent, id opaqueDataNext) {
+            mtr_hide(self); // don't capture self accidentally
             NSMutableArray<NSArray *> * readRequestsCurrent = opaqueDataCurrent;
             NSMutableArray<NSArray *> * readRequestsNext = opaqueDataNext;
 
-            // Can only read up to 9 paths at a time, per spec
-            if (readRequestsCurrent.count >= 9) {
-                MTR_LOG_DEFAULT("%@ batching cannot add more", logPrefix);
-                return;
-            }
-
+            MTRBatchingOutcome outcome = MTRNotBatched;
             while (readRequestsNext.count) {
+                // Can only read up to 9 paths at a time, per spec
+                if (readRequestsCurrent.count >= 9) {
+                    MTR_LOG_INFO("Batching cannot add more work, current item is full");
+                    return outcome;
+                }
+
                 // if params don't match then they cannot be merged
                 if (![readRequestsNext[0][MTRDeviceReadRequestFieldParamsIndex]
                         isEqual:readRequestsCurrent[0][MTRDeviceReadRequestFieldParamsIndex]]) {
-                    MTR_LOG_DEFAULT("%@ batching merged all possible items", logPrefix);
-                    return;
+                    MTR_LOG_INFO("Batching cannot add more work, parameter mismatch");
```

Comment:
Again, if we can do `add more work to item %tu`, that would be nice.  Or %llu if we change the type of the counter, of course.